mindmeld/active_learning/heuristics.py

```diff
@@ -0,0 +1,490 @@
+        Args:
+            confidences_2d (List[List[float]]): Confidence probabilities per element.
+        Returns:
+            ranked_indices (List[int]): Indices corresponding to elements ranked by utility.
+        """
+        highest_confidence_per_element = np.max(confidences_2d, axis=1)
+        return list(np.argsort(highest_confidence_per_element))
+
+    @staticmethod
+    def rank_3d(confidences_3d: List[List[List[float]]]) -> List[int]:
+        """First calculates the highest (max) confidences per element and then returns
+        the elements with the lowest max confidence. This is done for each confidence_2d
+        in a confidence_3d. The rankings are added to generate a final ranking.
+
+        Args:
+            confidences_3d (List[List[List[float]]]): Confidence probabilities per element.
+        Returns:
+            ranked_indices (List[int]): Indices corresponding to elements ranked by utility.
+        """
+        total_rank_score = np.sum(
+            [LeastConfidenceSampling.rank_2d(c) for c in confidences_3d], axis=0
+        )
+        high_to_low_ranked_indices = np.argsort(total_rank_score)[::-1]
+        return list(high_to_low_ranked_indices)
+
+
+class MarginSampling(ABC):
+    @staticmethod
+    def rank_2d(confidences_2d: List[List[float]]) -> List[int]:
+        """Calculates the "margin" or difference between the highest and second highest
+        confidence score per element. Elements are ranked from highest to lowest margin.
+
+        Args:
+            confidences_2d (List[List[float]]): Confidence probabilities per element.
+        Returns:
+            ranked_indices (List[int]): Indices corresponding to elements ranked by utility.
+        """
+        _, element_size = np.array(confidences_2d).shape
+        descending_confidences_per_element = np.partition(
+            confidences_2d, kth=(element_size - 1)
+        )[::-1]
+        highest_val_per_element = descending_confidences_per_element[:, 1]
+        second_highest_val_per_element = descending_confidences_per_element[:, 2]
+        margin_per_element = np.abs(
+            highest_val_per_element - second_highest_val_per_element
+        )
+        ranked_indices_high_to_low_margin = np.argsort(margin_per_element)[::-1]
+        return list(ranked_indices_high_to_low_margin)
+
+    @staticmethod
+    def rank_3d(confidences_3d: List[List[List[float]]]) -> List[int]:
+        """Calculates the "margin" or difference between the highest and second highest
+        confidence score per element. Elements are ranked from highest to lowest margin.
+        This is done for each confidence_2d in a confidence_3d. The rankings are added
+        to generate a final ranking.
+
+        Args:
+            confidences_3d (List[List[List[float]]]): Confidence probabilities per element.
+        Returns:
+            ranked_indices (List[int]): Indices corresponding to elements ranked by utility.
+        """
+        total_rank_score = np.sum(
```

Comment:
I think we might have the same problem as with least confidence sampling here ie, summing up sample indices rather than their ranks.

Reply:
Added a function to convert ordered indices to per-sample ranks. Modified this function accordingly